File: ScalaSelfType/src/com/bennyhuo/kotlin/generics/SelfType.scala
package com.bennyhuo.kotlin.generics

object Main {
    type OnConfirm = () => Unit
    type OnCancel = () => Unit

    final val EmptyFunction = () => {}

    class Notification(val title: String, val content: String)

    class ConfirmNotification(
                                 title: String,
                                 content: String,
                                 val onConfirm: OnConfirm,
                                 val onCancel: OnCancel
                             ) extends Notification(title, content)

    class NotificationBuilder {
        self =>
        protected var title: String = ""
        protected var content: String = ""

        def title(title: String): self.type = {
            this.title = title
            return this
        }

        def content(content: String): self.type = {
            this.content = content
            return this
        }

        def build() = new Notification(this.title, this.content)
    }

    class ConfirmNotificationBuilder extends NotificationBuilder() {
        private var onConfirm: OnConfirm = EmptyFunction
        private var onCancel: OnCancel = EmptyFunction

        def onConfirm(onConfirm: => Unit): ConfirmNotificationBuilder = {
            this.onConfirm = () => {
                onConfirm
            }
            return this
        }

        def onCancel(onCancel: => Unit): ConfirmNotificationBuilder = {
            this.onCancel = () => {
                onCancel
            }
            return this
        }

        override def build() = new ConfirmNotification(title, content, onConfirm, onCancel)
    }

    def main(args: Array[String]): Unit = {
        new ConfirmNotificationBuilder()
            .title("Hello")
            .onCancel {
                println("onCancel")
            }
            .content("World")
            .onConfirm {
                println("onConfirm")
            }
            .build()
            .onConfirm()
    }

}

File: Generics/src/main/java/com/bennyhuo/kotlin/generics/eg/SelfType.kt
package com.bennyhuo.kotlin.generics.eg

typealias OnConfirm = () -> Unit
typealias OnCancel = () -> Unit

private val EmptyFunction = {}

open class Notification(
    val title: String,
    val content: String
)

class ConfirmNotification(
    title: String,
    content: String,
    val onConfirm: OnConfirm,
    val onCancel: OnCancel
) : Notification(title, content)

interface SelfType<Self> {
    val self: Self
        get() = this as Self
}

open class NotificationBuilder<Self: NotificationBuilder<Self>>: SelfType<Self> {
    protected var title: String = ""
    protected var content: String = ""

    fun title(title: String): Self {
        this.title = title
        return self
    }

    fun content(content: String): Self {
        this.content = content
        return self
    }

    open fun build() = Notification(this.title, this.content)
}

class ConfirmNotificationBuilder : NotificationBuilder<ConfirmNotificationBuilder>() {
    private var onConfirm: OnConfirm = EmptyFunction
    private var onCancel: OnCancel = EmptyFunction

    fun onConfirm(onConfirm: OnConfirm): ConfirmNotificationBuilder {
        this.onConfirm = onConfirm
        return this
    }

    fun onCancel(onCancel: OnCancel): ConfirmNotificationBuilder {
        this.onCancel = onCancel
        return this
    }

    override fun build() = ConfirmNotification(title, content, onConfirm, onCancel)
}

fun main() {
    ConfirmNotificationBuilder()
        .title("Hello")
        .onCancel {
            println("onCancel")
        }.content("World")
        .onConfirm {
            println("onConfirmed")
        }
        .build()
        .onConfirm()
}
